src/bijectors/corr.jl

```diff
@@ -75,51 +75,158 @@ end
+
+function update_triu_from_vec!(
+    vals::AbstractVector{<:Real},
+    k::Int,
+    X::AbstractMatrix{<:Real}
+)
+    # Ensure that we're working with one-based indexing.
+    # `triu` requires this too.
+    LinearAlgebra.require_one_based_indexing(X)
+
+    # Set the values.
+    idx = 1
+    m, n = size(X)
+    for j = 1:n
+        for i = 1:min(j - k, m)
+            X[i, j] = vals[idx]
+            idx += 1
         end
     end
-    
-    return w
+
+    return X
+end
+
+function update_triu_from_vec(vals::AbstractVector{<:Real}, k::Int, dim::Int)
+    X = similar(vals, dim, dim)
+    # TODO: Do we need this?
+    X .= 0
+    return update_triu_from_vec!(vals, k, X)
+end
+
+function ChainRulesCore.rrule(::typeof(update_triu_from_vec), x::AbstractVector{<:Real}, k::Int, dim::Int)
+    function update_triu_from_vec_pullback(ΔX)
+        return (
+            ChainRulesCore.NoTangent(),
+            triu_to_vec(ChainRulesCore.unthunk(ΔX), k),
+            ChainRulesCore.NoTangent(),
+            ChainRulesCore.NoTangent()
+        )
+    end
+    return update_triu_from_vec(x, k, dim), update_triu_from_vec_pullback
+end
+
+#      n * (n - 1) / 2 = d
+# ⟺       n^2 - n - 2d = 0
+# ⟹                  n = (1 + sqrt(1 + 8d)) / 2
+_triu1_dim_from_length(d) = (1 + isqrt(1 + 8d)) ÷ 2
+
+"""
+    triu1_to_vec(X::AbstractMatrix{<:Real})
+
+Extracts elements from upper triangle of `X` with offset `1` and returns them as a vector.
+"""
+triu1_to_vec(X::AbstractMatrix) = triu_to_vec(X, 1)
+
+inverse(::typeof(triu1_to_vec)) = vec_to_triu1
+
+"""
+    vec_to_triu1(x::AbstractVector{<:Real})
+
+Constructs a matrix from a vector `x` by filling the upper triangle with offset `1`.
+"""
+function vec_to_triu1(x::AbstractVector)
+    n = _triu1_dim_from_length(length(x))
+    X = update_triu_from_vec(x, 1, n)
+    return upper_triangular(X)
+end
+
+inverse(::typeof(vec_to_triu1)) = triu1_to_vec
+
+"""
+    VecCorrBijector <: Bijector
+
+Similar to `CorrBijector`, but correlation matrix to a vector,
+and its inverse transforms vector to a correlation matrix.
+
+See also: [`CorrBijector`](@ref)
+
+# Example
+
+```jldoctest
+julia> using LinearAlgebra
+
+julia> using StableRNGs; rng = StableRNG(42);
+
+julia> b = Bijectors.VecCorrBijector();
+
+julia> X = rand(rng, LKJ(3, 1))  # Sample a correlation matrix.
+3×3 Matrix{Float64}:
+  1.0       -0.705273   -0.348638
+ -0.705273   1.0         0.0534538
+ -0.348638   0.0534538   1.0
+
+julia> y = b(X)  # Transform to unconstrained vector representation.
+3-element Vector{Float64}:
+ -0.8777149781928181
+ -0.3638927608636788
+ -0.29813769428942216
+
+julia> inverse(b)(y) ≈ X  # (✓) Round-trip through `b` and its inverse.
+true
+"""
+struct VecCorrBijector <: Bijector end
+with_logabsdet_jacobian(b::VecCorrBijector, x) = transform(b, x), logabsdetjac(b, x)
+
+function transform(::VecCorrBijector, X::AbstractMatrix{<:Real})
+    w = upper_triangular(parent(cholesky(X).U))
```

Comment:
I hate this:(

Reply:
Could we not have `w = cholesky(X).U` and work with a `w::UpperTriangular` instead of a dense matrix? Tried it locally, no real gain in performance though.

Reply:
> Could we not have w = cholesky(X).U and work with a w::UpperTriangular instead of a dense matrix? 

Yeah, but this won't work with some of the AD backends (I know, it's super-annoying...). If you have a look at our compat-code for ReverseDiff (I believe), I think you'll see that we have to do some custom stuff to compute the pullback.

> Tried it locally, no real gain in performance though.

I don't think we'd expect it to because internally we're iterating over the relevant elements of the matrix anyways, i.e. we're not gaining anything by telling the rest of the computational path that we're actually working on a lower-triangular matrix because it already assumes the given matrix is lower-triangular.

Reply:
I would guess it's mainly for efficiency reasons. But it's difficult to say if there are other implications as well, e.g., regarding AD. An alternative to `.UL` would be something like what's used in PDMats: https://github.com/JuliaStats/PDMats.jl/blob/fff131e11e23403931a42f5bfb3384f0d2b114c9/src/chol.jl#L6-L11 That should also be quite efficient and you could continue working with upper-triangular matrices.

Reply:
> I would guess it's mainly for efficiency reasons. 

https://github.com/JuliaLang/julia/blob/db7971f49912d1abba703345ca6eb43249607f32/stdlib/LinearAlgebra/src/cholesky.jl#L515-L527

> But it's difficult to say if there are other implications as well, e.g., regarding AD. 

Hmm fair, though IMO something like this seems like it would be a bug with the AD package, no?

> An alternative to .UL would be something like what's used in PDMats: https://github.com/JuliaStats/PDMats.jl/blob/fff131e11e23403931a42f5bfb3384f0d2b114c9/src/chol.jl#L6-L11 That should also be quite efficient and you could continue working with upper-triangular matrices.

Me and @harisorgn were just having a chat and we're thinking of replacing `upper_triangular(parent(cholesky(X).U))` with

```julia
cholesky_upper(x) = upper_triangular(parent(cholesky(X).U))
cholesky_lower(x) = lower_triangular(parent(cholesky(X).L))
```

to make it less likely that we forget or mess up somewhere.

But we can make it

```julia
cholesky_upper(x) = upper_triangular(parent(PDMats.chol_upper(cholesky(X))))
cholesky_lower(x) = lower_triangular(parent(PDMats.chol_lower(cholesky(X))))
```

But are you sure there's not a good reason for why the default is `copy`? Of course it's more mem-intensive, but will stuff like `LowerTriangular(U')` lead to slower computation paths (since you're now working with `adjoint(U)` rather than something that is actually lower-triangular)? E.g. indexing `adjoin(U)` surely involves more computations than indexing `copy(adjoint(U))`.